File: apina-core/src/main/java/fi/evident/apina/output/ts/CodeWriter.java
package fi.evident.apina.output.ts;

import java.util.Collection;
import java.util.Iterator;
import java.util.Map;

/**
 * Helper for generating TypeScript code. Keeps track of indentation level
 * and supports proper writing of literal values.
 */
final class CodeWriter {

    private final StringBuilder out = new StringBuilder();
    private int indentationLevel = 0;
    private boolean beginningOfLine = true;

    public CodeWriter writeLine(String s) {
        write(s);
        writeLine();
        return this;
    }

    public String getOutput() {
        return out.toString();
    }

    public CodeWriter write(String s) {
        if (!s.isEmpty()) {
            writeIndentIfAtBegin();
            out.append(s);
        }

        return this;
    }

    private void writeIndentIfAtBegin() {
        if (beginningOfLine) {
            writeIndent();
            beginningOfLine = false;
        }
    }

    public CodeWriter writeLine() {
        out.append('\n');
        beginningOfLine = true;
        return this;
    }

    public CodeWriter writeValue(Object obj) {
        writeIndentIfAtBegin();

        if (obj instanceof Number) {
            out.append(obj.toString());
        } else if (obj instanceof Map<?, ?>) {
            writeMap((Map<?, ?>) obj);
        } else if (obj instanceof String) {
            writeString((String) obj);
        } else if (obj instanceof Collection<?>) {
            writeCollection((Collection<?>) obj);
        } else {
            out.append(String.valueOf(obj));
        }

        return this;
    }

    public CodeWriter writeExportedModule(String module, Runnable moduleWriter) {
        return writeBlock("export module " + module, moduleWriter);
    }

    public CodeWriter writeExportedInterface(String name, Runnable moduleWriter) {
        return writeBlock("export interface " + name, moduleWriter);
    }

    public CodeWriter writeBlock(String prefix, Runnable moduleWriter) {
        return write(prefix + " ").writeBlock(moduleWriter).writeLine().writeLine();
    }

    public CodeWriter writeBlock(Runnable block) {
        writeLine("{");
        indent();

        block.run();

        dedent();
        write("}");

        return this;
    }

    private void writeMap(Map<?, ?> obj) {
        if (obj.isEmpty()) {
            write("{}");
            return;
        }

        writeBlock(() -> {
            for (Iterator<? extends Map.Entry<?, ?>> it = obj.entrySet().iterator(); it.hasNext(); ) {
                Map.Entry<?, ?> entry = it.next();
                writeValue(entry.getKey());
                write(": ");
                writeValue(entry.getValue());

                if (it.hasNext())
                    write(", ");

                writeLine();
            }
        });
    }

    private void writeCollection(Collection<?> obj) {
        write("[");

        for (Iterator<?> it = obj.iterator(); it.hasNext(); ) {
            writeValue(it.next());

            if (it.hasNext())
                write(", ");
        }

        write("]");
    }

    private void writeString(String s) {
        out.append('\'');
        for (int i = 0, len = s.length(); i < len; i++) {
            char c = s.charAt(i);
            switch (c) {
                case '\'':
                    out.append("\\'");
                    break;
                case '\n':
                    out.append("\\n");
                    break;
                default:
                    out.append(c);
            }
        }
        out.append('\'');
    }

    public CodeWriter indent() {
        indentationLevel++;
        return this;
    }

    public CodeWriter dedent() {
        if (indentationLevel == 0) throw new IllegalStateException();

        indentationLevel--;
        return this;
    }

    private void writeIndent() {
        for (int i = 0; i < indentationLevel; i++)
            out.append("    ");
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/output/ts/CodeWriter.kt
package fi.evident.apina.output.ts

/**
 * Helper for generating TypeScript code. Keeps track of indentation level
 * and supports proper writing of literal values.
 */
internal class CodeWriter {

    private val out = StringBuilder()
    private var indentationLevel = 0
    private var beginningOfLine = true

    fun writeLine(s: String): CodeWriter {
        write(s)
        writeLine()
        return this
    }

    val output: String
        get() = out.toString()

    private fun write(c: Char): CodeWriter {
        // Write indent if necessary, unless we are writing an empty line
        if (c != '\n' && beginningOfLine)
            out.append("    ".repeat(indentationLevel))

        out.append(c)

        beginningOfLine = c == '\n'

        return this
    }

    fun write(s: String): CodeWriter {
        for (c in s)
            write(c)

        return this
    }

    fun writeLine(): CodeWriter {
        write("\n")
        return this
    }

    fun writeValue(obj: Any?): CodeWriter {
        when (obj) {
            is Number ->
                write(obj.toString())
            is Map<*, *> ->
                writeMap(obj)
            is String ->
                writeString(obj)
            is Collection<*> ->
                writeCollection(obj)
            else ->
                write(obj.toString())
        }

        return this
    }

    fun writeExportedInterface(name: String, bodyWriter: () -> Unit): CodeWriter {
        return writeBlock("export interface $name", bodyWriter)
    }

    fun writeExportedClass(name: String, bodyWriter: () -> Unit): CodeWriter {
        return writeBlock("export class $name", bodyWriter)
    }

    fun writeBlock(prefix: String, bodyWriter: () -> Unit): CodeWriter {
        return write("$prefix ").writeBlock(bodyWriter).writeLine().writeLine()
    }

    fun writeImport(module: String, types: Collection<String>): CodeWriter {
        return writeLine("import { ${types.joinToString(", ")} } from '$module';")
    }

    fun writeBlock(block: () -> Unit): CodeWriter {
        writeLine("{")
        indent()

        block()

        dedent()
        write("}")

        return this
    }

    private fun writeMap(obj: Map<*, *>) {
        if (obj.isEmpty()) {
            write("{}")
            return
        }

        writeBlock {
            val it = obj.entries.iterator()
            while (it.hasNext()) {
                val entry = it.next()
                writeValue(entry.key)
                write(": ")
                writeValue(entry.value)

                if (it.hasNext())
                    write(",")

                writeLine()
            }
        }
    }

    private fun writeCollection(obj: Collection<*>) {
        write("[")

        val it = obj.iterator()
        while (it.hasNext()) {
            writeValue(it.next())

            if (it.hasNext())
                write(", ")
        }

        write("]")
    }

    private fun writeString(s: String) {
        write('\'')
        var i = 0
        val len = s.length
        while (i < len) {
            val c = s[i]
            when (c) {
                '\'' -> write("\\'")
                '\n' -> write("\\n")
                else -> write(c)
            }
            i++
        }
        write('\'')
    }

    private fun indent(): CodeWriter {
        indentationLevel++
        return this
    }

    private fun dedent(): CodeWriter {
        if (indentationLevel == 0) throw IllegalStateException()

        indentationLevel--
        return this
    }
}
